##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
  <td>
    <a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png" />See TF Hub models</a>
  </td>
</table>

This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with [TensorFlow Hub](https://tfhub.dev) and Keras.

It uses the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [`tf.keras`](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [`tensorflow_hub`](https://www.tensorflow.org/hub), a library for loading trained models from [TFHub](https://tfhub.dev) in a single line of code. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

In [3]:
!pip install tensorflow-hub
!pip install tensorflow-datasets

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.list_physical_devices("GPU") else "NOT AVAILABLE")

2023-12-07 03:16:12.593606: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-12-07 03:16:12.593651: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-12-07 03:16:12.595269: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


Version:  2.15.0
Eager mode:  True
Hub version:  0.15.0
GPU is available


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

In [5]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

Let's also print the first 10 labels.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. Use a pre-trained text embedding as the first layer, which will have three advantages:

*   You don't have to worry about text preprocessing,
*   Benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example you use a **pre-trained text embedding model** from [TensorFlow Hub](https://tfhub.dev) called [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2).

There are many other pre-trained text embeddings from TFHub that can be used in this tutorial:

* [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - trained with the same NNLM architecture on the same data as [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2), but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model.
* [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - the same as [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2), but with additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation.
* [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - a much larger model yielding 512 dimensional embeddings trained with a deep averaging network (DAN) encoder.

And many more! Find more [text embedding models](https://tfhub.dev/s?module-type=text-embedding) on TFHub.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423194 , -0.01190171,  0.06337537,  0.0686297 , -0.16776839,
        -0.10581177,  0.168653  , -0.04998823, -0.31148052,  0.07910344,
         0.15442258,  0.01488661,  0.03930155,  0.19772716, -0.12215477,
        -0.04120982, -0.27041087, -0.21922147,  0.26517656, -0.80739075,
         0.25833526, -0.31004202,  0.2868321 ,  0.19433866, -0.29036498,
         0.0386285 , -0.78444123, -0.04793238,  0.41102988, -0.36388886,
        -0.58034706,  0.30269453,  0.36308962, -0.15227163, -0.4439151 ,
         0.19462997,  0.19528405,  0.05666233,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201319 , -0.04418665, -0.08550781,
        -0.55847436, -0.2333639 , -0.20782956, -0.03543065, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862677,  0.7753425 , -0.07667087,
        -0.15752274,  0.01872334, -0.08169781, -0.3521876 ,  0.46373403,
        -0.08492758,  0.07166861, -0.00670818,  0.12686071, -0.19326551,
 

Let's now build the full model:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"


_________________________________________________________________


 Layer (type)                Output Shape              Param #   


 keras_layer (KerasLayer)    (None, 50)                48190600  


 dense (Dense)               (None, 16)                816       


 dense_1 (Dense)             (None, 1)                 17        


Total params: 48191433 (183.84 MB)


Trainable params: 48191433 (183.84 MB)


Non-trainable params: 0 (0.00 Byte)


_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that you are using ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`. For this NNLM model, the `embedding_dimension` is 50.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs logits (a single-unit layer with a linear activation), you'll use the `binary_crossentropy` loss function.

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when you are exploring regression problems (say, to predict the price of a house), you'll see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## Train the model

Train the model for 10 epochs in mini-batches of 512 samples. This is 10 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


I0000 00:00:1701918985.657311  104571 device_compiler.h:186] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


 1/30 [>.............................] - ETA: 1:11 - loss: 0.6724 - accuracy: 0.5625

 2/30 [=>............................] - ETA: 5s - loss: 0.6631 - accuracy: 0.5684  

 3/30 [==>...........................] - ETA: 4s - loss: 0.6667 - accuracy: 0.5625

 4/30 [===>..........................] - ETA: 4s - loss: 0.6647 - accuracy: 0.5483

 5/30 [====>.........................] - ETA: 4s - loss: 0.6616 - accuracy: 0.5414

 6/30 [=====>........................] - ETA: 4s - loss: 0.6605 - accuracy: 0.5378

 7/30 [======>.......................] - ETA: 4s - loss: 0.6601 - accuracy: 0.5377

 8/30 [=======>......................] - ETA: 4s - loss: 0.6584 - accuracy: 0.5391

 9/30 [========>.....................] - ETA: 3s - loss: 0.6567 - accuracy: 0.5334

10/30 [=========>....................] - ETA: 3s - loss: 0.6556 - accuracy: 0.5311

11/30 [==========>...................] - ETA: 3s - loss: 0.6536 - accuracy: 0.5330

12/30 [===========>..................] - ETA: 3s - loss: 0.6518 - accuracy: 0.5322

13/30 [============>.................] - ETA: 3s - loss: 0.6503 - accuracy: 0.5322

14/30 [=============>................] - ETA: 2s - loss: 0.6481 - accuracy: 0.5328

15/30 [==============>...............] - ETA: 2s - loss: 0.6463 - accuracy: 0.5349

16/30 [===============>..............] - ETA: 2s - loss: 0.6435 - accuracy: 0.5388

17/30 [================>.............] - ETA: 2s - loss: 0.6417 - accuracy: 0.5427

18/30 [=================>............] - ETA: 2s - loss: 0.6406 - accuracy: 0.5445

19/30 [==================>...........] - ETA: 2s - loss: 0.6384 - accuracy: 0.5479

20/30 [===================>..........] - ETA: 1s - loss: 0.6365 - accuracy: 0.5512

21/30 [====================>.........] - ETA: 1s - loss: 0.6352 - accuracy: 0.5543

22/30 [=====================>........] - ETA: 1s - loss: 0.6332 - accuracy: 0.5569

23/30 [======================>.......] - ETA: 1s - loss: 0.6315 - accuracy: 0.5592

24/30 [=======================>......] - ETA: 1s - loss: 0.6295 - accuracy: 0.5616

25/30 [========================>.....] - ETA: 0s - loss: 0.6282 - accuracy: 0.5630

26/30 [=========================>....] - ETA: 0s - loss: 0.6264 - accuracy: 0.5663

27/30 [==========================>...] - ETA: 0s - loss: 0.6241 - accuracy: 0.5689

28/30 [===========================>..] - ETA: 0s - loss: 0.6224 - accuracy: 0.5712

29/30 [============================>.] - ETA: 0s - loss: 0.6207 - accuracy: 0.5735

30/30 [==============================] - ETA: 0s - loss: 0.6203 - accuracy: 0.5744

30/30 [==============================] - 9s 213ms/step - loss: 0.6203 - accuracy: 0.5744 - val_loss: 0.5576 - val_accuracy: 0.6664


Epoch 2/10


 1/30 [>.............................] - ETA: 12:02 - loss: 0.5391 - accuracy: 0.6777

 2/30 [=>............................] - ETA: 5s - loss: 0.5326 - accuracy: 0.6904   

 3/30 [==>...........................] - ETA: 4s - loss: 0.5339 - accuracy: 0.6979

 4/30 [===>..........................] - ETA: 4s - loss: 0.5313 - accuracy: 0.7041

 5/30 [====>.........................] - ETA: 4s - loss: 0.5332 - accuracy: 0.7102

 6/30 [=====>........................] - ETA: 4s - loss: 0.5296 - accuracy: 0.7204

 7/30 [======>.......................] - ETA: 4s - loss: 0.5270 - accuracy: 0.7227

 8/30 [=======>......................] - ETA: 4s - loss: 0.5265 - accuracy: 0.7251

 9/30 [========>.....................] - ETA: 3s - loss: 0.5265 - accuracy: 0.7222

10/30 [=========>....................] - ETA: 3s - loss: 0.5242 - accuracy: 0.7240

11/30 [==========>...................] - ETA: 3s - loss: 0.5237 - accuracy: 0.7221

12/30 [===========>..................] - ETA: 3s - loss: 0.5220 - accuracy: 0.7222

13/30 [============>.................] - ETA: 3s - loss: 0.5200 - accuracy: 0.7206

14/30 [=============>................] - ETA: 2s - loss: 0.5184 - accuracy: 0.7203

15/30 [==============>...............] - ETA: 2s - loss: 0.5176 - accuracy: 0.7195

16/30 [===============>..............] - ETA: 2s - loss: 0.5164 - accuracy: 0.7208

17/30 [================>.............] - ETA: 2s - loss: 0.5138 - accuracy: 0.7241

18/30 [=================>............] - ETA: 2s - loss: 0.5113 - accuracy: 0.7276

20/30 [===================>..........] - ETA: 1s - loss: 0.5059 - accuracy: 0.7369

21/30 [====================>.........] - ETA: 1s - loss: 0.5039 - accuracy: 0.7388

22/30 [=====================>........] - ETA: 1s - loss: 0.5004 - accuracy: 0.7410

23/30 [======================>.......] - ETA: 1s - loss: 0.4983 - accuracy: 0.7426

24/30 [=======================>......] - ETA: 1s - loss: 0.4973 - accuracy: 0.7433

25/30 [========================>.....] - ETA: 0s - loss: 0.4944 - accuracy: 0.7455

26/30 [=========================>....] - ETA: 0s - loss: 0.4924 - accuracy: 0.7473

27/30 [==========================>...] - ETA: 0s - loss: 0.4897 - accuracy: 0.7495

28/30 [===========================>..] - ETA: 0s - loss: 0.4877 - accuracy: 0.7508

29/30 [============================>.] - ETA: 0s - loss: 0.4863 - accuracy: 0.7509

30/30 [==============================] - ETA: 0s - loss: 0.4857 - accuracy: 0.7514

30/30 [==============================] - 31s 208ms/step - loss: 0.4857 - accuracy: 0.7514 - val_loss: 0.4508 - val_accuracy: 0.7807


Epoch 3/10


 1/30 [>.............................] - ETA: 46s - loss: 0.3925 - accuracy: 0.8477

 2/30 [=>............................] - ETA: 5s - loss: 0.4054 - accuracy: 0.8389 

 3/30 [==>...........................] - ETA: 5s - loss: 0.4016 - accuracy: 0.8281

 4/30 [===>..........................] - ETA: 4s - loss: 0.3964 - accuracy: 0.8281

 5/30 [====>.........................] - ETA: 4s - loss: 0.3962 - accuracy: 0.8281

 6/30 [=====>........................] - ETA: 4s - loss: 0.3949 - accuracy: 0.8288

 7/30 [======>.......................] - ETA: 4s - loss: 0.3952 - accuracy: 0.8295

 8/30 [=======>......................] - ETA: 4s - loss: 0.3910 - accuracy: 0.8345

 9/30 [========>.....................] - ETA: 3s - loss: 0.3886 - accuracy: 0.8344

11/30 [==========>...................] - ETA: 3s - loss: 0.3859 - accuracy: 0.8361

12/30 [===========>..................] - ETA: 3s - loss: 0.3834 - accuracy: 0.8363

13/30 [============>.................] - ETA: 2s - loss: 0.3820 - accuracy: 0.8364

14/30 [=============>................] - ETA: 2s - loss: 0.3816 - accuracy: 0.8368

16/30 [===============>..............] - ETA: 2s - loss: 0.3791 - accuracy: 0.8378

17/30 [================>.............] - ETA: 2s - loss: 0.3776 - accuracy: 0.8395

18/30 [=================>............] - ETA: 1s - loss: 0.3761 - accuracy: 0.8401

19/30 [==================>...........] - ETA: 1s - loss: 0.3730 - accuracy: 0.8419

20/30 [===================>..........] - ETA: 1s - loss: 0.3705 - accuracy: 0.8435

21/30 [====================>.........] - ETA: 1s - loss: 0.3681 - accuracy: 0.8450

22/30 [=====================>........] - ETA: 1s - loss: 0.3663 - accuracy: 0.8461

23/30 [======================>.......] - ETA: 1s - loss: 0.3648 - accuracy: 0.8469

25/30 [========================>.....] - ETA: 0s - loss: 0.3612 - accuracy: 0.8484

26/30 [=========================>....] - ETA: 0s - loss: 0.3597 - accuracy: 0.8492

27/30 [==========================>...] - ETA: 0s - loss: 0.3591 - accuracy: 0.8498

28/30 [===========================>..] - ETA: 0s - loss: 0.3574 - accuracy: 0.8511

29/30 [============================>.] - ETA: 0s - loss: 0.3558 - accuracy: 0.8527

30/30 [==============================] - ETA: 0s - loss: 0.3556 - accuracy: 0.8531

30/30 [==============================] - 27s 869ms/step - loss: 0.3556 - accuracy: 0.8531 - val_loss: 0.3718 - val_accuracy: 0.8284


Epoch 4/10


 1/30 [>.............................] - ETA: 19s - loss: 0.2906 - accuracy: 0.8867

 2/30 [=>............................] - ETA: 5s - loss: 0.2868 - accuracy: 0.8896 

 4/30 [===>..........................] - ETA: 3s - loss: 0.2864 - accuracy: 0.8872

 5/30 [====>.........................] - ETA: 3s - loss: 0.2868 - accuracy: 0.8887

 6/30 [=====>........................] - ETA: 3s - loss: 0.2862 - accuracy: 0.8867

 7/30 [======>.......................] - ETA: 3s - loss: 0.2842 - accuracy: 0.8873

 8/30 [=======>......................] - ETA: 3s - loss: 0.2829 - accuracy: 0.8879

 9/30 [========>.....................] - ETA: 3s - loss: 0.2824 - accuracy: 0.8900

10/30 [=========>....................] - ETA: 3s - loss: 0.2812 - accuracy: 0.8936

11/30 [==========>...................] - ETA: 3s - loss: 0.2786 - accuracy: 0.8961

12/30 [===========>..................] - ETA: 3s - loss: 0.2766 - accuracy: 0.8973

13/30 [============>.................] - ETA: 2s - loss: 0.2753 - accuracy: 0.8972

14/30 [=============>................] - ETA: 2s - loss: 0.2738 - accuracy: 0.8977

15/30 [==============>...............] - ETA: 2s - loss: 0.2724 - accuracy: 0.8977

16/30 [===============>..............] - ETA: 2s - loss: 0.2701 - accuracy: 0.8982

17/30 [================>.............] - ETA: 2s - loss: 0.2700 - accuracy: 0.8982

18/30 [=================>............] - ETA: 2s - loss: 0.2685 - accuracy: 0.8988

19/30 [==================>...........] - ETA: 1s - loss: 0.2683 - accuracy: 0.8983

20/30 [===================>..........] - ETA: 1s - loss: 0.2672 - accuracy: 0.8987

21/30 [====================>.........] - ETA: 1s - loss: 0.2664 - accuracy: 0.8998

22/30 [=====================>........] - ETA: 1s - loss: 0.2652 - accuracy: 0.9007

23/30 [======================>.......] - ETA: 1s - loss: 0.2643 - accuracy: 0.9008

24/30 [=======================>......] - ETA: 1s - loss: 0.2644 - accuracy: 0.9006

25/30 [========================>.....] - ETA: 0s - loss: 0.2633 - accuracy: 0.9009

27/30 [==========================>...] - ETA: 0s - loss: 0.2618 - accuracy: 0.9004

28/30 [===========================>..] - ETA: 0s - loss: 0.2609 - accuracy: 0.9007

29/30 [============================>.] - ETA: 0s - loss: 0.2594 - accuracy: 0.9011

30/30 [==============================] - ETA: 0s - loss: 0.2594 - accuracy: 0.9011

30/30 [==============================] - 7s 218ms/step - loss: 0.2594 - accuracy: 0.9011 - val_loss: 0.3327 - val_accuracy: 0.8588


Epoch 5/10


 1/30 [>.............................] - ETA: 26s - loss: 0.2137 - accuracy: 0.9453

 2/30 [=>............................] - ETA: 5s - loss: 0.2129 - accuracy: 0.9375 

 3/30 [==>...........................] - ETA: 4s - loss: 0.2101 - accuracy: 0.9362

 4/30 [===>..........................] - ETA: 4s - loss: 0.2061 - accuracy: 0.9390

 5/30 [====>.........................] - ETA: 4s - loss: 0.2041 - accuracy: 0.9402

 6/30 [=====>........................] - ETA: 4s - loss: 0.2022 - accuracy: 0.9421

 7/30 [======>.......................] - ETA: 4s - loss: 0.2000 - accuracy: 0.9411

 8/30 [=======>......................] - ETA: 4s - loss: 0.2015 - accuracy: 0.9402

 9/30 [========>.....................] - ETA: 3s - loss: 0.2015 - accuracy: 0.9390

10/30 [=========>....................] - ETA: 3s - loss: 0.2007 - accuracy: 0.9391

11/30 [==========>...................] - ETA: 3s - loss: 0.1992 - accuracy: 0.9380

12/30 [===========>..................] - ETA: 3s - loss: 0.1973 - accuracy: 0.9383

13/30 [============>.................] - ETA: 3s - loss: 0.1980 - accuracy: 0.9369

14/30 [=============>................] - ETA: 2s - loss: 0.1980 - accuracy: 0.9361

15/30 [==============>...............] - ETA: 2s - loss: 0.1990 - accuracy: 0.9353

16/30 [===============>..............] - ETA: 2s - loss: 0.1980 - accuracy: 0.9355

17/30 [================>.............] - ETA: 2s - loss: 0.1970 - accuracy: 0.9354

18/30 [=================>............] - ETA: 2s - loss: 0.1955 - accuracy: 0.9368

19/30 [==================>...........] - ETA: 2s - loss: 0.1963 - accuracy: 0.9364

20/30 [===================>..........] - ETA: 1s - loss: 0.1956 - accuracy: 0.9368

21/30 [====================>.........] - ETA: 1s - loss: 0.1951 - accuracy: 0.9370

23/30 [======================>.......] - ETA: 1s - loss: 0.1950 - accuracy: 0.9358

24/30 [=======================>......] - ETA: 1s - loss: 0.1947 - accuracy: 0.9355

25/30 [========================>.....] - ETA: 0s - loss: 0.1939 - accuracy: 0.9356

26/30 [=========================>....] - ETA: 0s - loss: 0.1936 - accuracy: 0.9352

27/30 [==========================>...] - ETA: 0s - loss: 0.1931 - accuracy: 0.9357

28/30 [===========================>..] - ETA: 0s - loss: 0.1925 - accuracy: 0.9360

29/30 [============================>.] - ETA: 0s - loss: 0.1924 - accuracy: 0.9360

30/30 [==============================] - ETA: 0s - loss: 0.1920 - accuracy: 0.9361

30/30 [==============================] - 7s 207ms/step - loss: 0.1920 - accuracy: 0.9361 - val_loss: 0.3139 - val_accuracy: 0.8648


Epoch 6/10


 1/30 [>.............................] - ETA: 21s - loss: 0.1415 - accuracy: 0.9727

 2/30 [=>............................] - ETA: 5s - loss: 0.1458 - accuracy: 0.9619 

 3/30 [==>...........................] - ETA: 4s - loss: 0.1470 - accuracy: 0.9603

 4/30 [===>..........................] - ETA: 4s - loss: 0.1440 - accuracy: 0.9629

 5/30 [====>.........................] - ETA: 4s - loss: 0.1444 - accuracy: 0.9613

 6/30 [=====>........................] - ETA: 4s - loss: 0.1444 - accuracy: 0.9606

 7/30 [======>.......................] - ETA: 4s - loss: 0.1454 - accuracy: 0.9587

 8/30 [=======>......................] - ETA: 4s - loss: 0.1487 - accuracy: 0.9556

 9/30 [========>.....................] - ETA: 3s - loss: 0.1485 - accuracy: 0.9559

10/30 [=========>....................] - ETA: 3s - loss: 0.1495 - accuracy: 0.9553

11/30 [==========>...................] - ETA: 3s - loss: 0.1508 - accuracy: 0.9551

12/30 [===========>..................] - ETA: 3s - loss: 0.1505 - accuracy: 0.9551

13/30 [============>.................] - ETA: 3s - loss: 0.1475 - accuracy: 0.9563

14/30 [=============>................] - ETA: 2s - loss: 0.1471 - accuracy: 0.9562

15/30 [==============>...............] - ETA: 2s - loss: 0.1472 - accuracy: 0.9559

16/30 [===============>..............] - ETA: 2s - loss: 0.1475 - accuracy: 0.9562

17/30 [================>.............] - ETA: 2s - loss: 0.1470 - accuracy: 0.9562

18/30 [=================>............] - ETA: 2s - loss: 0.1457 - accuracy: 0.9568

19/30 [==================>...........] - ETA: 2s - loss: 0.1449 - accuracy: 0.9568

20/30 [===================>..........] - ETA: 1s - loss: 0.1451 - accuracy: 0.9563

21/30 [====================>.........] - ETA: 1s - loss: 0.1451 - accuracy: 0.9561

22/30 [=====================>........] - ETA: 1s - loss: 0.1447 - accuracy: 0.9565

23/30 [======================>.......] - ETA: 1s - loss: 0.1447 - accuracy: 0.9562

25/30 [========================>.....] - ETA: 0s - loss: 0.1438 - accuracy: 0.9571

26/30 [=========================>....] - ETA: 0s - loss: 0.1428 - accuracy: 0.9575

27/30 [==========================>...] - ETA: 0s - loss: 0.1428 - accuracy: 0.9578

28/30 [===========================>..] - ETA: 0s - loss: 0.1423 - accuracy: 0.9579

29/30 [============================>.] - ETA: 0s - loss: 0.1421 - accuracy: 0.9577

30/30 [==============================] - ETA: 0s - loss: 0.1423 - accuracy: 0.9575

30/30 [==============================] - 7s 204ms/step - loss: 0.1423 - accuracy: 0.9575 - val_loss: 0.3079 - val_accuracy: 0.8640


Epoch 7/10


 1/30 [>.............................] - ETA: 19s - loss: 0.1074 - accuracy: 0.9727

 2/30 [=>............................] - ETA: 5s - loss: 0.1129 - accuracy: 0.9658 

 4/30 [===>..........................] - ETA: 3s - loss: 0.1152 - accuracy: 0.9658

 5/30 [====>.........................] - ETA: 3s - loss: 0.1198 - accuracy: 0.9676

 6/30 [=====>........................] - ETA: 3s - loss: 0.1181 - accuracy: 0.9697

 7/30 [======>.......................] - ETA: 3s - loss: 0.1171 - accuracy: 0.9701

 8/30 [=======>......................] - ETA: 3s - loss: 0.1150 - accuracy: 0.9707

 9/30 [========>.....................] - ETA: 3s - loss: 0.1130 - accuracy: 0.9716

10/30 [=========>....................] - ETA: 3s - loss: 0.1140 - accuracy: 0.9711

11/30 [==========>...................] - ETA: 3s - loss: 0.1136 - accuracy: 0.9712

12/30 [===========>..................] - ETA: 3s - loss: 0.1122 - accuracy: 0.9717

14/30 [=============>................] - ETA: 2s - loss: 0.1105 - accuracy: 0.9717

15/30 [==============>...............] - ETA: 2s - loss: 0.1108 - accuracy: 0.9714

16/30 [===============>..............] - ETA: 2s - loss: 0.1093 - accuracy: 0.9717

17/30 [================>.............] - ETA: 2s - loss: 0.1077 - accuracy: 0.9724

18/30 [=================>............] - ETA: 1s - loss: 0.1072 - accuracy: 0.9727

19/30 [==================>...........] - ETA: 1s - loss: 0.1067 - accuracy: 0.9730

20/30 [===================>..........] - ETA: 1s - loss: 0.1061 - accuracy: 0.9732

21/30 [====================>.........] - ETA: 1s - loss: 0.1060 - accuracy: 0.9735

23/30 [======================>.......] - ETA: 1s - loss: 0.1060 - accuracy: 0.9740

24/30 [=======================>......] - ETA: 0s - loss: 0.1052 - accuracy: 0.9738

25/30 [========================>.....] - ETA: 0s - loss: 0.1057 - accuracy: 0.9731

26/30 [=========================>....] - ETA: 0s - loss: 0.1053 - accuracy: 0.9727

27/30 [==========================>...] - ETA: 0s - loss: 0.1056 - accuracy: 0.9718

28/30 [===========================>..] - ETA: 0s - loss: 0.1052 - accuracy: 0.9720

29/30 [============================>.] - ETA: 0s - loss: 0.1050 - accuracy: 0.9721

30/30 [==============================] - ETA: 0s - loss: 0.1046 - accuracy: 0.9723

30/30 [==============================] - 6s 196ms/step - loss: 0.1046 - accuracy: 0.9723 - val_loss: 0.3100 - val_accuracy: 0.8702


Epoch 8/10


 1/30 [>.............................] - ETA: 21s - loss: 0.0893 - accuracy: 0.9863

 2/30 [=>............................] - ETA: 5s - loss: 0.0907 - accuracy: 0.9854 

 3/30 [==>...........................] - ETA: 4s - loss: 0.0914 - accuracy: 0.9844

 5/30 [====>.........................] - ETA: 3s - loss: 0.0830 - accuracy: 0.9863

 6/30 [=====>........................] - ETA: 3s - loss: 0.0810 - accuracy: 0.9854

 7/30 [======>.......................] - ETA: 3s - loss: 0.0792 - accuracy: 0.9855

 9/30 [========>.....................] - ETA: 3s - loss: 0.0771 - accuracy: 0.9848

10/30 [=========>....................] - ETA: 2s - loss: 0.0770 - accuracy: 0.9850

11/30 [==========>...................] - ETA: 2s - loss: 0.0767 - accuracy: 0.9851

12/30 [===========>..................] - ETA: 2s - loss: 0.0769 - accuracy: 0.9845

14/30 [=============>................] - ETA: 2s - loss: 0.0765 - accuracy: 0.9838

15/30 [==============>...............] - ETA: 2s - loss: 0.0766 - accuracy: 0.9833

16/30 [===============>..............] - ETA: 2s - loss: 0.0764 - accuracy: 0.9828

17/30 [================>.............] - ETA: 2s - loss: 0.0761 - accuracy: 0.9828

19/30 [==================>...........] - ETA: 1s - loss: 0.0760 - accuracy: 0.9826

21/30 [====================>.........] - ETA: 1s - loss: 0.0752 - accuracy: 0.9833

22/30 [=====================>........] - ETA: 1s - loss: 0.0757 - accuracy: 0.9829

24/30 [=======================>......] - ETA: 0s - loss: 0.0759 - accuracy: 0.9825

25/30 [========================>.....] - ETA: 0s - loss: 0.0756 - accuracy: 0.9824

27/30 [==========================>...] - ETA: 0s - loss: 0.0751 - accuracy: 0.9824

28/30 [===========================>..] - ETA: 0s - loss: 0.0746 - accuracy: 0.9827

30/30 [==============================] - ETA: 0s - loss: 0.0752 - accuracy: 0.9821

30/30 [==============================] - 5s 164ms/step - loss: 0.0752 - accuracy: 0.9821 - val_loss: 0.3170 - val_accuracy: 0.8675


Epoch 9/10


 1/30 [>.............................] - ETA: 24s - loss: 0.0557 - accuracy: 0.9883

 2/30 [=>............................] - ETA: 5s - loss: 0.0547 - accuracy: 0.9893 

 3/30 [==>...........................] - ETA: 5s - loss: 0.0574 - accuracy: 0.9883

 4/30 [===>..........................] - ETA: 4s - loss: 0.0567 - accuracy: 0.9883

 6/30 [=====>........................] - ETA: 3s - loss: 0.0560 - accuracy: 0.9886

 7/30 [======>.......................] - ETA: 3s - loss: 0.0567 - accuracy: 0.9891

 8/30 [=======>......................] - ETA: 3s - loss: 0.0563 - accuracy: 0.9897

 9/30 [========>.....................] - ETA: 3s - loss: 0.0563 - accuracy: 0.9902

10/30 [=========>....................] - ETA: 3s - loss: 0.0574 - accuracy: 0.9900

11/30 [==========>...................] - ETA: 3s - loss: 0.0568 - accuracy: 0.9902

12/30 [===========>..................] - ETA: 3s - loss: 0.0552 - accuracy: 0.9907

13/30 [============>.................] - ETA: 2s - loss: 0.0554 - accuracy: 0.9907

14/30 [=============>................] - ETA: 2s - loss: 0.0558 - accuracy: 0.9909

16/30 [===============>..............] - ETA: 2s - loss: 0.0553 - accuracy: 0.9910

17/30 [================>.............] - ETA: 2s - loss: 0.0547 - accuracy: 0.9910

19/30 [==================>...........] - ETA: 1s - loss: 0.0543 - accuracy: 0.9907

21/30 [====================>.........] - ETA: 1s - loss: 0.0549 - accuracy: 0.9899

22/30 [=====================>........] - ETA: 1s - loss: 0.0550 - accuracy: 0.9901

23/30 [======================>.......] - ETA: 1s - loss: 0.0546 - accuracy: 0.9901

24/30 [=======================>......] - ETA: 0s - loss: 0.0546 - accuracy: 0.9900

25/30 [========================>.....] - ETA: 0s - loss: 0.0550 - accuracy: 0.9898

26/30 [=========================>....] - ETA: 0s - loss: 0.0550 - accuracy: 0.9897

27/30 [==========================>...] - ETA: 0s - loss: 0.0545 - accuracy: 0.9899

28/30 [===========================>..] - ETA: 0s - loss: 0.0541 - accuracy: 0.9901

29/30 [============================>.] - ETA: 0s - loss: 0.0538 - accuracy: 0.9902

30/30 [==============================] - ETA: 0s - loss: 0.0537 - accuracy: 0.9903

30/30 [==============================] - 6s 186ms/step - loss: 0.0537 - accuracy: 0.9903 - val_loss: 0.3281 - val_accuracy: 0.8687


Epoch 10/10


 1/30 [>.............................] - ETA: 17s - loss: 0.0485 - accuracy: 0.9922

 2/30 [=>............................] - ETA: 5s - loss: 0.0470 - accuracy: 0.9902 

 3/30 [==>...........................] - ETA: 5s - loss: 0.0474 - accuracy: 0.9902

 5/30 [====>.........................] - ETA: 3s - loss: 0.0443 - accuracy: 0.9934

 6/30 [=====>........................] - ETA: 3s - loss: 0.0419 - accuracy: 0.9941

 8/30 [=======>......................] - ETA: 3s - loss: 0.0423 - accuracy: 0.9937

 9/30 [========>.....................] - ETA: 3s - loss: 0.0422 - accuracy: 0.9941

10/30 [=========>....................] - ETA: 3s - loss: 0.0423 - accuracy: 0.9943

12/30 [===========>..................] - ETA: 2s - loss: 0.0413 - accuracy: 0.9943

14/30 [=============>................] - ETA: 2s - loss: 0.0404 - accuracy: 0.9947

15/30 [==============>...............] - ETA: 2s - loss: 0.0403 - accuracy: 0.9948

16/30 [===============>..............] - ETA: 2s - loss: 0.0402 - accuracy: 0.9946

17/30 [================>.............] - ETA: 1s - loss: 0.0401 - accuracy: 0.9943

19/30 [==================>...........] - ETA: 1s - loss: 0.0397 - accuracy: 0.9941

20/30 [===================>..........] - ETA: 1s - loss: 0.0395 - accuracy: 0.9943

21/30 [====================>.........] - ETA: 1s - loss: 0.0398 - accuracy: 0.9941

22/30 [=====================>........] - ETA: 1s - loss: 0.0395 - accuracy: 0.9941

23/30 [======================>.......] - ETA: 1s - loss: 0.0392 - accuracy: 0.9943

24/30 [=======================>......] - ETA: 0s - loss: 0.0392 - accuracy: 0.9943

25/30 [========================>.....] - ETA: 0s - loss: 0.0390 - accuracy: 0.9942

26/30 [=========================>....] - ETA: 0s - loss: 0.0390 - accuracy: 0.9941

27/30 [==========================>...] - ETA: 0s - loss: 0.0386 - accuracy: 0.9944

28/30 [===========================>..] - ETA: 0s - loss: 0.0386 - accuracy: 0.9944

29/30 [============================>.] - ETA: 0s - loss: 0.0386 - accuracy: 0.9943

30/30 [==============================] - ETA: 0s - loss: 0.0385 - accuracy: 0.9944

30/30 [==============================] - 6s 184ms/step - loss: 0.0385 - accuracy: 0.9944 - val_loss: 0.3415 - val_accuracy: 0.8669


## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3715 - accuracy: 0.8503 - 1s/epoch - 30ms/step


loss: 0.371
accuracy: 0.850


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

* For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, see the [Text classification with preprocessed text](./text_classification.ipynb) tutorial.
* Try out more [text-related tutorials](https://www.tensorflow.org/hub/tutorials#text-related-tutorials) using trained models from TFHub.